### PR TITLE
New matching task queue ownership for task queue ownership control

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Add task queue ownership management logic
  * Add task ID management
  * Add task flushing wrapper
  * Add UT

**Why?**
Matching task queue logic cleanup

**How did you test it?**
New tests

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

File: service/matching/db_task_ownership.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+)
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination db_task_ownership_mock.go
+
+const (
+	dbTaskOwnershipStatusUninitialized dbTaskOwnershipStatus = 0
+	dbTaskOwnershipStatusOwning        dbTaskOwnershipStatus = 1
```
Comment (paulnpdev) [inline service/matching/db_task_ownership.go:46]: Owned better than Owning here

File: service/matching/db_task_ownership.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+		getLastAllocatedTaskID() int64
+		persistTaskQueue() error
+		flushTasks(taskInfos ...*persistencespb.TaskInfo) error
+	}
+
+	dbTaskOwnershipState struct {
+		rangeID             int64
+		ackedTaskID         int64
+		lastAllocatedTaskID int64
+		minTaskID           int64 // exclusive
```
Comment (paulnpdev) [inline service/matching/db_task_ownership.go:66]: maybe make the inclusivity part of the name so user doesn't have to come here to look it up (i.e. minTaskIdExclusive)?

File: service/matching/db_task_ownership.go [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+		taskQueueKind   enumspb.TaskQueueKind
+		taskIDRangeSize int64
+		timeSource      clock.TimeSource
+		store           persistence.TaskManager
+		logger          log.Logger
+
+		sync.Mutex
+		status              dbTaskOwnershipStatus
+		ownershipState      *dbTaskOwnershipState
+		shutdownC           chan struct{}
```
Comment (paulnpdev) [inline service/matching/db_task_ownership.go:81]: shutdownChan?

File: service/matching/db_task_ownership.go [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+		stateLastUpdateTime *time.Time
+	}
+)
+
+func newDBTaskOwnership(
```
Comment (paulnpdev) [inline service/matching/db_task_ownership.go:86]: isn't this "logically public"?  (i.e. NewDBTaskOwnership)?
your call

File: service/matching/db_task_ownership.go [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+
+		status:              dbTaskOwnershipStatusUninitialized,
+		ownershipState:      nil,
+		shutdownC:           make(chan struct{}),
+		stateLastUpdateTime: nil,
+	}
+	return taskOwnership
+}
+
+func (m *dbTaskOwnershipImpl) shutdownChan() <-chan struct{} {
```
Comment (paulnpdev) [inline service/matching/db_task_ownership.go:109]: getShutdownChan() would be more consistent ...

**Review by paulnpdev — APPROVED** [review]

File: service/matching/db_task_ownership.go (hunk 4 above)
Comment (wxing1292) [reply service/matching/db_task_ownership.go:86]: this entity will only be used by the matching package (it is already in the same package)
so private, not public

**Pull request merged**